### PR TITLE
Masking configuration values irrelevant to DAG author

### PR DESCRIPTION
**Pull request by amoghrajesh**
Some configurations are irrelevant to DAG authors and hence we need to mask those to avoid it from getting logged unknowingly.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/settings.py [hunk 1]
```diff
@@ -722,6 +743,16 @@ def import_local_settings():
         log.info("Loaded airflow_local_settings from %s .", airflow_local_settings.__file__)
 
 
+def mask_conf_values():
+    from airflow.utils.log.secrets_masker import mask_secret
+
+    for env in ENV_VARIABLES_TO_MASK:
+        section, key = conf.get_section_and_key_for_env(env)
```
Comment (amoghrajesh) [inline airflow/settings.py]: The reason this is done is so that if the config is set in any other way than env, we can detect it and mask it

**Comment by amoghrajesh** [comment]
Struggling with some test related setup. I am running into this as of now:
```
../../tests_common/test_utils/db.py:23: in <module>
    from airflow.models import (
../../airflow/models/__init__.py:78: in __getattr__
    val = import_string(f"{path}.{name}")
../../airflow/utils/module_loading.py:39: in import_string
    module = import_module(module_path)
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
../../airflow/models/dag.py:95: in <module>
    from airflow.models.abstractoperator import AbstractOperator, TaskStateChangeCallback
../../airflow/models/abstractoperator.py:33: in <module>
    from airflow.template.templater import Templater
../../airflow/template/templater.py:23: in <module>
    from airflow.io.path import ObjectStoragePath
../../airflow/io/__init__.py:30: in <module>
    from airflow.providers_manager import ProvidersManager
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    """Manages all providers."""
    
    from __future__ import annotations
    
    import fnmatch
    import functools
    import inspect
    import json
    import logging
    import os
    import sys
    import traceback
    import warnings
    from dataclasses import dataclass
    from functools import wraps
    from time import perf_counter
    from typing import TYPE_CHECKING, Any, Callable, MutableMapping, NamedTuple, TypeVar
    
    from packaging.utils import canonicalize_name
    
    from airflow.exceptions import AirflowOptionalProviderFeatureException
>   from airflow.providers.standard.hooks.filesystem import FSHook
E   ModuleNotFoundError: No module named 'airflow.providers.standard.hooks'
```

I see some discussions around, does anyone have any suggestions for this? I use Pycharm

**Comment by potiuk** [comment]
> I see some discussions around, does anyone have any suggestions for this? I use Pycharm

I believe (I have not yet had time to look at it) this requires to use `uv workspace` and the latest `uv` installed - @ashb - are there any instructions for that already somewhere ? 

I guess those should be updated ?

* https://github.com/apache/airflow/blob/main/contributing-docs/07_local_virtualenv.rst
* https://github.com/apache/airflow/blob/main/contributing-docs/quick-start-ide/contributors_quick_start_pycharm.rst
* https://github.com/apache/airflow/blob/main/contributing-docs/quick-start-ide/contributors_quick_start_vscode.rst

**Comment by potiuk** [comment]
I think we generally should provide rather comprehensive guide (at least with some links) to our contributors how to setup the venv now (Breeze is covered because it uses workspace installation internally I believe)? Am I right @ashb ?

File: airflow/settings.py [hunk 2]
```diff
@@ -118,6 +118,27 @@
     "upstream_failed": "orange",
 }
 
+# list of the environment variables that need to be masked from the logs
+# some of these could be sensitive and there's no reason to log those
+ENV_VARIABLES_TO_MASK = [
+    "AIRFLOW__CORE__FERNET_KEY",
+    "AIRFLOW__SMTP__SMTP_PORT",
+    "AIRFLOW__SMTP__SMTP_PASSWORD",
```
Comment (ashb) [inline airflow/settings.py]: This list feels wrong to hard code. I'm fairly sure there is already a concept in the configuration of a "sensitive" config which we should reuse.

File: tests/core/test_settings.py [hunk 3]
```diff
@@ -385,3 +390,21 @@ def test_usage_data_collection_disabled(env_var, conf_setting, is_enabled, clear
...
         with conf_patch:
             assert is_usage_data_collection_enabled() == is_enabled
+
+
+@patch("airflow.utils.log.secret_masker.mask_secret")
+@patch("airflow.configuration.conf.get")
+@patch("airflow.configuration.get_section_and_key_for_env")
+@patch("airflow.settings.ENV_VARIABLES_TO_MASK", ["AIRFLOW__CORE__FERNET_KEY", "AIRFLOW__NONMASK_ENV"])
+def test_mask_conf_values(self, mock_get_section_and_key, mock_get, mock_mask_secret):
+    mock_get_section_and_key.side_effect = [
```
Comment (ashb) [inline tests/core/test_settings.py]: I don't understand why this needs to be mocked - by mocking it you never test the function you added

File: airflow/configuration.py [hunk 4]
```diff
@@ -775,6 +775,13 @@ def _create_future_warning(name: str, section: str, current_value: Any, new_valu
     def _env_var_name(self, section: str, key: str) -> str:
         return f"{ENV_VAR_PREFIX}{section.replace('.', '_').upper()}__{key.upper()}"
 
+    def get_section_and_key_for_env(self, env: str):
```
Comment (ashb) [inline airflow/configuration.py]: I don't get why this is needed. If we need a list (again, we already have a concept of sensitive config values I think) then we should specify the list items as `(section,key)` etc

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by ashb** [comment]
Breeze is all set up yes. UV workspace isn't 100% required yet (though I don't think anything tested running tests from inside pycharm 0, but for pycharm to find imports you need to set up some paths as the right kind of folder in the UI 

I'm not a pycharm user, but @kaxil @dstandish can provide some insight.

File: airflow/settings.py (hunk 2 above)
Comment (kaxil) [reply airflow/settings.py]: Yeah, the following ones: 
- https://github.com/apache/airflow/blob/aaa5f3566663eb3e9a0e7bf0522d533fbb1010d5/airflow/config_templates/config.yml#L182
- https://github.com/apache/airflow/blob/aaa5f3566663eb3e9a0e7bf0522d533fbb1010d5/airflow/configuration.py#L297-L323

Comment (ashb) [reply airflow/settings.py]: Also. Masking the PORT? Why?!

File: airflow/settings.py [hunk 5]
```diff
@@ -118,6 +118,27 @@
     "upstream_failed": "orange",
 }
 
+# list of the environment variables that need to be masked from the logs
+# some of these could be sensitive and there's no reason to log those
+ENV_VARIABLES_TO_MASK = [
+    "AIRFLOW__CORE__FERNET_KEY",
+    "AIRFLOW__SMTP__SMTP_PORT",
```
Comment (kaxil) [inline airflow/settings.py]: This isn't sensitive

File: airflow/settings.py (hunk 2 above)
Comment (kaxil) [inline airflow/settings.py]: Smtp password was already removed: https://github.com/apache/airflow/pull/41539

File: airflow/settings.py [hunk 6]
```diff
@@ -118,6 +118,27 @@
...
+    "AIRFLOW__SMTP__SMTP_PASSWORD",
+    "AIRFLOW__SMTP__SMTP_PASSWORD_SECRET",
```
Comment (ashb) [inline airflow/settings.py]: The `_secret` aren't sensitive either -- they are the name of a secret to fetch the creds from so aren't sensitive themselves.

**Comment by ashb** [comment]
https://github.com/apache/airflow/pull/42951 should help Pycharm too.

File: airflow/settings.py (hunk 2 above)
Comment (amoghrajesh) [reply airflow/settings.py]: Well yeah that makes sense, let me consume those and try to mask those.

**Comment by dstandish** [comment]
> I'm not a pycharm user, but @kaxil @dstandish can provide some insight.

Anything in particular you wanted me to look at?

**Comment by ashb** [comment]
> Should these be masked or its alright to keep them unmasked?

https://github.com/apache/airflow/pull/43040#discussion_r1802935540

**Comment by amoghrajesh** [comment]
@potiuk @ashb added a unit test here. Can you take a look at this when you have some time?

File: airflow/settings.py [hunk 7]
```diff
@@ -722,6 +722,22 @@ def import_local_settings():
...
+        try:
+            value = conf.get(section, key)
+        except AirflowConfigException:
+            log.warning(
+                "Could not retrieve value from section %s, for key %s. Skipping redaction of this conf.",
+                section,
+                key,
+            )
+            continue
+        mask_secret(value)
```
Comment (ashb) [inline airflow/settings.py]: This feels like it should be method on Configuration, not a few method in settings module.

Could you move it and the tests please?

(Calling conf.mask_secrets from withing settings is still okay I think)

Comment (amoghrajesh) [reply airflow/settings.py:738]: I cannot call the mask_secret from configuration. It gets a circular import there

Comment (amoghrajesh) [reply airflow/settings.py:738]: Okay let me do that

Comment (amoghrajesh) [reply airflow/settings.py]: @ashb the challenge in doing that would be importing the `mask_secret` inside configuration.py.
It would lead to a circular import.

**Comment by amoghrajesh** [comment]
Hey @ashb. Left a comment on your last review, can you please take a look when you have some time?

File: airflow/settings.py (hunk 7 above)
Comment (ashb) [reply airflow/settings.py]: Ah right k. Well we could choose to import `mask_secret` only inside this function. I guess it's fine here where it is

File: airflow/settings.py [hunk 8]
```diff
@@ -722,6 +722,22 @@ def import_local_settings():
...
+    for section, key in conf.sensitive_config_values:
+        try:
+            value = conf.get(section, key)
+        except AirflowConfigException:
+            log.warning(
+                "Could not retrieve value from section %s, for key %s. Skipping redaction of this conf.",
+                section,
+                key,
+            )
+            continue
```
Comment (ashb) [inline airflow/settings.py]: How many messages is this going to create every time Airflow starts up? This would also appear inside the task instance logs, so I think this should probably be reduced to a debug level log so that it doesn't "spam" things.

File: airflow/settings.py (hunk 7 above)
Comment (ashb) [reply airflow/settings.py]: Wait, the import is already inside this function anyway, so it wouldn't create a circular import (only top level imports can create circular imports. (Circular dependency, sure, but that's a fact of life in python really.)

Comment (amoghrajesh) [reply airflow/settings.py]: Alright, slightly confused with your comment. I have imported mask_secret only in here. Should I make some changes or this is alright?

**Review by ashb — APPROVED** [review]
Left my final comment about log level. Preemptively approving assuming that change is made.

Since the import is already inside the function (i.e. no circular imports caused) please also move it to live in configuration.py

File: airflow/settings.py (hunk 7 above)
Comment (ashb) [reply airflow/settings.py]: I'm saying, that since the import is here, that we can move this function to be a method on our Configuration class in airflow/configuration.py without causing any import cycles. Importing things inside a function to avoid import loops is a pattern we do a lot in Airflow.

File: airflow/settings.py (hunk 8 above)
Comment (amoghrajesh) [reply airflow/settings.py]: Makes sense, this will log a couple of times.
I will make this change and move it to DEBUG so that the log flooding is much lesser.

File: airflow/settings.py (hunk 7 above)
Comment (amoghrajesh) [reply airflow/settings.py]: Let me try that out, makes sense.

**Comment by amoghrajesh** [comment]
@ashb you are right, it didn't work the last time and I landed at circular imports due to a top level import. 
It works now, pushed the new changes. Still assuming your +1 as I handled your review comments.

@potiuk / @kaxil could you take a look when you have some time?

File: airflow/configuration.py [hunk 9]
```diff
@@ -772,6 +772,21 @@ def _create_future_warning(name: str, section: str, current_value: Any, new_valu
             stacklevel=3,
         )
 
+    def mask_secrets(self):
+        from airflow.utils.log.secrets_masker import mask_secret
+
+        for section, key in conf.sensitive_config_values:
+            try:
+                value = conf.get(section, key)
```
Comment (ashb) [inline airflow/configuration.py]:
```suggestion
        for section, key in self.sensitive_config_values:
            try:
                value = self.get(section, key)
```

File: tests/core/test_configuration.py [hunk 10]
```diff
@@ -1763,3 +1763,22 @@ def test_config_paths_is_directory(self):
...
+        new_callable=lambda: [("mysection1", "mykey1"), ("mysection2", "mykey2")],
+    )
+    @patch("airflow.utils.log.secrets_masker.mask_secret")
+    @patch("airflow.configuration.conf.get")
+    def test_mask_conf_values(self, mock_get, mock_mask_secret, mock_sensitive_config_values):
+        mock_get.side_effect = ["supersecret1", "supersecret2"]
+        conf.mask_secrets()
+
+        mock_get.assert_any_call("mysection1", "mykey1")
+        mock_get.assert_any_call("mysection2", "mykey2")
```
Comment (ashb) [inline tests/core/test_configuration.py]: Nit, rather than mocking `conf.get` please use the existing `@conf_vars` decorator.

File: tests/core/test_settings.py [hunk 11]
```diff
@@ -31,7 +31,11 @@
 from airflow.api_internal.internal_api_call import InternalApiConfig
 from airflow.configuration import conf
 from airflow.exceptions import AirflowClusterPolicyViolation, AirflowConfigException
-from airflow.settings import _ENABLE_AIP_44, TracebackSession, is_usage_data_collection_enabled
+from airflow.settings import (
+    _ENABLE_AIP_44,
+    TracebackSession,
+    is_usage_data_collection_enabled,
+)
```
Comment (ashb) [inline tests/core/test_settings.py]: Nit: unrelated changes in this file now, lets revert these ones

```suggestion
from airflow.settings import _ENABLE_AIP_44, TracebackSession, is_usage_data_collection_enabled
```

File: tests/core/test_configuration.py (hunk 10 above)
Comment (amoghrajesh) [reply tests/core/test_configuration.py]: Sure, made the changes in latest commit

**Comment by amoghrajesh** [comment]
@ashb does it look good to go? Got a green CI

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Looks like all the comments are resolved and address by @amoghrajesh . LGTM. @kaxil @ashb - are you ok with merging it?

**Review by ashb — APPROVED** [review]

**Comment by kaxil** [comment]
This change has caused the `@suppress_logs_and_warning` decorator to not work for CLI commands and hence CLI commands now show warnings. Example: https://github.com/apache/airflow/pull/43334

**Comment by kaxil** [comment]
Fix here: https://github.com/apache/airflow/pull/43335

**Comment by saurabhb-dev** [comment]
@amoghrajesh , shouldn't the list of sensitive_config_values be updated with the list of keys that need to redact ? 
Line: https://github.com/amoghrajesh/airflow/blob/840ea3efb9533837e9f36b75fa527a0fbafeb23a/airflow/configuration.py#L121

**Comment by potiuk** [comment]
> @amoghrajesh , shouldn't the list of sensitive_config_values be updated with the list of keys that need to redact ? Line: https://github.com/amoghrajesh/airflow/blob/840ea3efb9533837e9f36b75fa527a0fbafeb23a/airflow/configuration.py#L121

Not really @saurabhb-dev , It's even better to mask values in this case, rather than keys.

The secrets_masker works in two modes:

* for known structures it can mask sensitive key values (for example if a key in connection is "password" - it will mask the value when it prints the structure (when it is aware of the structure and acts on it)

* but this is impossible when you log messages when the structures or values have been already converted to log string - by the time secrets masker in logger (installed as filter) gets the message, the structure is gone already (converted to string representation) and we are not aware that particular key corresponds to particular value. So our secrets masker has the second mode - where it can mask specific values. It will scan the string for all the values that are registered upfront and mask them if it finds any of them in the string.

This is what happens here - we retrieve all the secret config values (whether they come by env vars or by other means) and we add values (i.e. actual secrets) to be masked. This way when any secret is printed anywhere where secrets_masker is used, it will automatically mask it - regardless if it is a structure (secrets_masker checks values of dicts for example) or whether it's already converted to string.

**Comment by zachliu** [comment]
figure i'd ask here before creating a new issue

https://github.com/apache/airflow/blob/c99887ec11ce3e1a43f2794fcf36d27555140f00/airflow/configuration.py#L857-L859

`self.sensitive_config_values` returns both `('database', 'sql_alchemy_conn')` and `('core', 'sql_alchemy_conn')`

hence the `self.get()` raises a warning:

```
airflow/configuration.py:863 FutureWarning: section/key [core/sql_alchemy_conn] has been deprecated, you should use[database/sql_alchemy_conn] instead. Please update your `conf.get*` call to use the new name
```

which is a bit confusing since i removed all references to `core.sql_alchemy_conn` long ago

**Comment by amoghrajesh** [comment]
@zachliu we do not return ('database', 'sql_alchemy_conn') as part of self.sensitive_config_values. Here are the sensitive values: https://github.com/amoghrajesh/airflow/blob/840ea3efb9533837e9f36b75fa527a0fbafeb23a/airflow/configuration.py#L122-L129

The warning comes from configuration.py#919

**Comment by zachliu** [comment]
@amoghrajesh i beg to differ

1. the warning says `configuration.py:859`
2. i tested with a print statement at these locations:
    * `configuration.py:859`
    * `configuration.py:323`


i'm on https://github.com/apache/airflow/tree/2.10.3

**Comment by zachliu** [comment]
i think i know what's going on here. somehow https://github.com/apache/airflow/commit/7dea23fd23fa347143fded6979879cf5f58fa132 (https://github.com/apache/airflow/pull/42126) didn't make the cut for 2.10.3, which caused the `self.sensitive_config_values` to return both `('database', 'sql_alchemy_conn')` and `('core', 'sql_alchemy_conn')`

**Comment by zachliu** [comment]
I created a new issue to track this: https://github.com/apache/airflow/issues/43794
